beacon_node/network/src/sync/block_lookups/common.rs

```diff
@@ -0,0 +1,448 @@
+    fn lookup_type() -> LookupType {
+        LookupType::Current
+    }
+}
+
+/// This trait unifies common single block lookup functionality across blocks and blobs. This
+/// includes making requests, verifying responses, and handling processing results. A
+/// `SingleBlockLookup` includes both a `BlockRequestState` and a `BlobRequestState`, this trait is
+/// implemented for each.
+///
+/// The use of the `ResponseType` associated type gives us a degree of type
+/// safety when handling a block/blob response ensuring we only mutate the correct corresponding
+/// state.
+pub trait RequestState<L: Lookup, T: BeaconChainTypes> {
+    /// The type of the request .
+    type RequestType;
+
+    /// A block or blob response.
+    type ResponseType;
+
+    /// The type created after validation.
+    type VerifiedResponseType: Clone;
+
+    /// We convert a `VerifiedResponseType` to this type prior to sending it to the beacon processor.
+    type ReconstructedResponseType;
+
+    /* Request building methods */
+
+    /// Construct a new request.
+    fn build_request(&mut self) -> Result<(PeerId, Self::RequestType), LookupRequestError> {
+        debug_assert!(matches!(self.get_state().state, State::AwaitingDownload));
+        self.too_many_attempts()?;
+        let peer = self.get_peer()?;
+        let request = self.new_request();
+        self.get_state_mut().req_counter += 1;
+        Ok((peer, request))
+    }
+
+    /// Verify the current request has not exceeded the maximum number of attempts.
+    fn too_many_attempts(&self) -> Result<(), LookupRequestError> {
+        let max_attempts = L::max_attempts();
+        let request_state = self.get_state();
+
+        if request_state.failed_attempts() >= max_attempts {
+            let cannot_process =
+                request_state.failed_processing >= request_state.failed_downloading;
+            Err(LookupRequestError::TooManyAttempts { cannot_process })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Get the next peer to request. Draws from the set of peers we think should have both the
+    /// block and blob first. If that fails, we draw from the set of peers that may have either.
+    fn get_peer(&mut self) -> Result<PeerId, LookupRequestError> {
+        let request_state = self.get_state_mut();
+        let available_peer_opt = request_state
+            .available_peers
+            .iter()
+            .choose(&mut rand::thread_rng())
+            .copied()
+            .map(PeerShouldHave::BlockAndBlobs);
+
+        let Some(peer_id) = available_peer_opt.or_else(||request_state
+            .potential_peers
+            .iter()
+            .choose(&mut rand::thread_rng())
+            .copied()
+            .map(PeerShouldHave::Neither)) else {
+            return Err(LookupRequestError::NoPeers);
+        };
+        request_state.used_peers.insert(peer_id.to_peer_id());
+        request_state.state = State::Downloading { peer_id };
```

Comment:
I guess this assumes the implementer of the trait then actually sends the request. 

I can see the `make_request` is generic so I can see why its put here. Just a note to be careful that we could get out of state if implementations are not as expected.

Reply:
Maybe its worth documenting which functions need to call what in order for the implementation to be consistent.

Reply:
Would it be possible to make `make_request` a concrete function which updates the request state, and then calls the generic function `make_request_impl` (need better name 😂 )? Could be messy though, just a thought.

Reply:
Just realized `make_request` implementation is actually really simple, so my above suggestion is probably an overkill, feel free to ignore. I think what Age suggested makes sense.

Reply:
Ok I think I made the request building vs sending separation make more sense here:

74373a9bd7a460911fcc3e7e8b23018c2506f821